ANDROID: Add SAF support

lephilousophe opened the pull request:
This PR finally brings Storage Access Framework support to ScummVM.
This has only been tested on Android 10 phone for now.
Performances seemed decent.
I tried to cache as much things as I can to avoid querying over and over the SAF backend.

Many tests are needed.

As I created a real Android backend (based on posix-drives one), this allowed me to remove all the specific code in `gui` and `backends/fs/posix`.

I took the opportunity to fix some errors and upgrade Android build tools.
NDK has not been upgraded as it has several consequences on the toolchain.

bluegr commented on backends/fs/android/android-saf-fs.cpp:
```diff
@@ -0,0 +1,698 @@
+// which gets messed up by our override mechanism; this could
+// be avoided by either changing the Android SDK to use the equally
+// legal and valid
+//   __attribute__ ((format(printf, 3, 4)))
```
This line is the same as the one in line 31 above. A typo, perhaps?

lephilousophe replied:
It is but it's a comment I pasted from android.cpp.
Fixing for coherence everywhere.

bluegr commented on backends/fs/android/android-fs-factory.cpp:
```diff
@@ -0,0 +1,100 @@
+		_config.drives.push_back(path);
+	}
+
+	// For old POSIX way
```
This code branch runs even if `_withSAF` is false. Is this intentional?

lephilousophe replied:
Thanks, I added _withSAF on the late and missed a few spots.
SAF part should not be iterated as JNI parameters are not initialized.
POSIX part must run as it's the legacy drives I removed from `PosixFilesystemNode`

bluegr commented on backends/fs/android/android-saf-fs.cpp:
```diff
@@ -0,0 +1,698 @@
+}
+
+AbstractFSNode *AddSAFFakeNode::getChild(const Common::String &name) const {
+	// We can't call getChild as it's protected
```
Is this a TODO? Or is this implementation of `getChild()` not used by engine code?

lephilousophe replied:
This is a fake node which is only used in the browser at the root (you can't create this fake node using a path).
It's kind of a hack to allow adding SAF trees without having to modify the GUI.

Normally `getChild` isn't used by the browser because it already used `getChildren`.

I don't plan to change the AbstractFSNode API to make it not protected.

If a change would be needed, it means another way to add the SAF node must be found (don't add hacks over hacks).

bluegr commented on backends/platform/android/org/scummvm/scummvm/SAFFSTree.java:
```diff
@@ -0,0 +1,439 @@
+		return newnode;
+	}
+
+	public SAFFSNode createFile(SAFFSNode node, String name) {
```
Since this code is 99% the same as `createDirectory()` above, perhaps it would make sense to reuse it in an internal funcion?

lephilousophe replied:
Also done.

bluegr commented on backends/platform/android/org/scummvm/scummvm/SAFFSTree.java:
```diff
@@ -0,0 +1,439 @@
+		return pfd.detachFd();
+	}
+
+	public int createWriteStream(SAFFSNode node) {
```
Since this code is 99% the same as `createReadStream()` above, perhaps it would make sense to reuse it using an internal function?

lephilousophe replied:
Indeed. Done.

bluegr commented on backends/platform/android/org/scummvm/scummvm/ScummVMActivity.java:
```diff
@@ -2293,245 +2132,98 @@ private void copyAssetsToInternalMemory(boolean sideUpgrade) {
-		Uri treeUri;
-		if ((treeUri = getStorageAccessFrameworkTreeUri()) == null) {
-			return;
+		intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION | Intent.FLAG_GRANT_WRITE_URI_PERMISSION);
```
We no longer request intent `FLAG_GRANT_PERSISTABLE_URI_PERMISSION`. According to the [intent documentation](https://developer.android.com/reference/android/content/Intent#FLAG_GRANT_PERSISTABLE_URI_PERMISSION), "_When combined with [FLAG_GRANT_READ_URI_PERMISSION](https://developer.android.com/reference/android/content/Intent#FLAG_GRANT_READ_URI_PERMISSION) and/or [FLAG_GRANT_WRITE_URI_PERMISSION](https://developer.android.com/reference/android/content/Intent#FLAG_GRANT_WRITE_URI_PERMISSION), the URI permission grant can be persisted across device reboots until explicitly revoked with [Context#revokeUriPermission(Uri, int)](https://developer.android.com/reference/android/content/Context#revokeUriPermission(android.net.Uri,%20int))_"

Is this not needed anymore? Should it be a TODO?

lephilousophe replied:
After taking a look at documentation, Android source code and samples on the net, I believe these flags are useless in this way.
These flags are set in the result intent we get in the `onActivityResult` function and are not used when querying.

bluegr commented:
Very nice work! Making this an actual backend clears up the code considerably.
The new implementation is of very high standard and quality.

Can't wait to get this merged :) This will help us get ScummVM back to the Google Play Store.
Kudos! :)

lephilousophe commented:
@bluegr Thanks for your review. I addressed your comments.

After more testing, I found a crash with backtrace when quitting and restarting the app immediately.
I still have to check about the why.

ccawley2011 commented on backends/fs/android/android-saf-fs.cpp:
```diff
@@ -0,0 +1,698 @@
+		return nullptr;
+	}
+
+	return new StdioStream(f);
```
Would `PosixIoStream` work here as well?

lephilousophe replied:
It seems to work.
I tested to detect a game and it worked.

ccawley2011 commented on backends/fs/android/android-posix-fs.cpp:
```diff
@@ -0,0 +1,61 @@
+AndroidPOSIXFilesystemNode::AndroidPOSIXFilesystemNode(const Common::String &path,
+        const Config &config) :
+	DrivePOSIXFilesystemNode(path, config) {
+	if (((uintptr_t)&path >= (uintptr_t)config.drives.begin()) &&
```
Could you add a comment here explaining what this does?

lephilousophe replied:
Sure, done.

lephilousophe commented:
And the crash is gone. :)
I left the deletion of _fsFactory. This cannot be done now like that as it's a singleton.

lephilousophe commented:
OK, so I rewrited partially what I did by making DrivePOSIXFilesystemNode more generic.
Now drives are made by the Config class which is made generic.
I need to make a build with NDS to make sure it still builds but I am quite confident.

This let me remove the hack about String pointers to override display name of drives.
I also cleared cache when hiding the activity to take into account user actions when ScummVM was not used.

I will now try to add some GUI to remove SAF authorizations instead of clearing all of them at once.

lephilousophe commented:
SAF UI is done.
It's available from the backend tab in global options.

SAF is hidden on Android below Lollipop 5 which added support for OPEN_DOCUMENT_TREE (tested on KitKat 4.4 old phone).
On Android Lollipop 5 and up to Android Q (10) included, both modes are supported.
Starting with Android R (11), old mode doesn't work anymore.

A manual readd of games would be needed for users migrating from Android 10 to 11 because no migration path is available to map old SDCard paths to new SAF ones.
This is not planned.

I think I have finished my work here and this can be merged as soon as enough testing has been done.

A screenshot of the new removal UI:
![Screenshot_1673897497](https://user-images.githubusercontent.com/1492917/212754072-2e6a7d74-da2d-4207-9130-dc193811d1a7.png)
